### ISSUE TITLE
Middleware-Status for Koa 2.0 transition

### ISSUE THREAD

**Issue by dj-hedgehog** [strapi/strapi#41]
Koa replaces their generator-style middleware implementation for [async functions](https://github.com/tc39/ecmascript-asyncawait) starting in version 2.0. All middlewares used by Strapi need to adapt in order to work with version 2.0. This issue shall be used to track to progress of the transition.

**Please use this issue to comment if you have updated information or want to discuss the strapi's transition to the new promise-based middleware implementation**
## NOTES
- [koa-convert](https://github.com/gyson/koa-convert) could be used to convert most middleware from generator to promise-based middleware and vice versa
- koa tracks its status [here](https://github.com/koajs/koa/issues/603)
## General async function support
- [x] V8 includes native async-functions - [v8 Issue #4483](https://code.google.com/p/v8/issues/detail?id=4483)
- [x] node.js supports native async-functions
## Middleware support
- [x] [koa-bodyparser](https://github.com/koajs/bodyparser) — [Branch 3.x](https://github.com/koajs/bodyparser/tree/3.x)
- [x] [koa-compose](https://github.com/koajs/compose) — [Branch v2.x](https://github.com/koajs/compose/tree/next)
- [x] [koa-compress](https://github.com/koajs/compress) — [Branch v2.x](https://github.com/koajs/compress/tree/v2.x)
- [x] [kcors](https://github.com/koajs/cors/) — [Branch v2.x](https://github.com/koajs/cors/tree/v2.x)
- [x] [koa-favicon](https://github.com/koajs/favicon) —  [Branch v2.x](https://github.com/koajs/favicon/tree/v2.x)
- [x] [koa-i18n](https://github.com/koa-modules/i18n) — [Branch master](https://github.com/koa-modules/i18n)
- [x] [koa-locale](https://github.com/koa-modules/locale)
- [ ] [koa-lusca](https://github.com/koajs/koa-lusca)
- [x] [koa-mount](https://github.com/koajs/mount) — [Branch next](https://github.com/koajs/mount/tree/next)
- [x] [koa-proxy](https://github.com/popomore/koa-proxy)
- [x] [koa-response-time](https://github.com/koajs/response-time) — [Branch 2.x](https://github.com/koajs/response-time/tree/v2.x)
- [x] [joi-router](https://github.com/koajs/joi-router) — [Branch await](https://github.com/koajs/joi-router/tree/await)
- [x] [koa-send](https://github.com/koajs/send) — [Branch 2.x](https://github.com/koajs/send/tree/2.x)
- [x] [koa-generic-session](https://github.com/koajs/generic-session)
- [x] [koa-sslify](https://github.com/turboMaCk/koa-sslify) — [Branch koa2](https://github.com/turboMaCk/koa-sslify/tree/koa2)
- [ ] [koa-better-static](https://github.com/ohomer/koa-better-static)
- [x] [koa-views](https://github.com/queckezz/koa-views) — [Branch master](https://github.com/queckezz/koa-views)

**Comment by dj-hedgehog** [comment]
Updated `koa-send` and `koa-compose` as they now contain a v2 compatible branch.
Removed `koa-load-middleware` from the list as it needs no updating to be compatible with koa v2.

**Comment by fundon** [comment]
Released koa-i18n@v2.

**Comment by loicsaintroch** [comment]
Awesome! Thanks for the update @fundon we appreciate.

**Comment by dj-hedgehog** [comment]
I've exchanged `koa-gzip` with `koa-compress` here too (see https://github.com/wistityhq/strapi/pull/87) because we won't switch to koa v2 before Strapi 2.0 is released.

I also propose adding a `Future` milestone and adding this Issue to it.
The reason is there is currently no estimate on when async/await will land in V8 and even worse in Node itself. Therefore this would delay Strapi 2.0 greatly.

**Comment by loicsaintroch** [comment]
@dj-hedgehog I agree with you. However I prefer to change the milestone when we'll work with our new OSS workflow. I'll write some blog posts about the v2, our vision and our goals to be completely honest and transparent with the community.

**Comment by dj-hedgehog** [comment]
The situation keeps getting brighter koa-response-time, koa-static, koa-compress & koa-views all have respective branches for v2.x. I'll take care of some of the missing ones myself when I find the time.

**Comment by Aurelsicoko** [comment]
I suggest moving from `koa-ssl` to `koa-sslify`. `koa-sslify` looks more maintained and popular. The middleware offers more options such as Azure support and reverse proxies. We can't stay on an unused middleware without any update since 2 years.

**Comment by dj-hedgehog** [comment]
Now, things are getting interesting as there is native `async..await` support in the just released Node v7.
I suggest moving to Koa v2 now because that makes working with Strapi even easier. Also v3 isn't released yet.

**Comment by Aurelsicoko** [comment]
@dj-hedgehog `async/await` seems buggy on V8 5.4 and have some nasty memory leaks... I will create a branch `koa@next` and start moving to Koa v2. Furthermore, it requires `--harmony-async-await` flag to start. The `harmony` flag is not recommended for production use, though. However, the `strapi-v3@alpha` will not be ready for production use so it is not a real problem for now.

FYI, I'm following [this thread](https://github.com/koajs/koa/issues/533#issuecomment-256249601) on Koa's repository to get in touch.

**Comment by dj-hedgehog** [comment]
@Aurelsicoko I didn't know about the leaks. Will definitely test this some more and follow the thread you mentioned.

**Comment by Aurelsicoko** [comment]
I started my work on [this branch](https://github.com/strapi/strapi/tree/improvement/koa2) `improvement/koa2`. It works pretty well. It seems faster to load. Furthermore, we don't need generator functions into our controllers anymore.

**Comment by StevenACoffman** [comment]
by the way [koa-cors next branch](https://github.com/kunyan/koa-cors/tree/next)

**Comment by Aurelsicoko** [comment]
@StevenACoffman Thanks for the link! I was wondering if we should move to the official CORS's middleware for Koa (https://github.com/koajs/cors). It provides the exactly same features and this middleware is highly maintained by the Koa core team.

**Comment by StevenACoffman** [comment]
@Aurelsicoko Yeah, unless there is some sign of recent activity, I think you're probably right to switch to the official package. I hadn't realized the official project existed.

**Comment by Aurelsicoko** [comment]
@StevenACoffman Me too... I moved to the official package today on the branch `improvement/koa2` and it works fine!

**Comment by Aurelsicoko** [comment]
FYI, we are using [koa-generic-session](https://github.com/koajs/generic-session) instead of [koa-session](https://github.com/koajs/session). Also, we are no longer using `koa-graphql` by default. As I previously said, the GraphQL support will be available through a plugin.

Summary of changes we made:
- [koa-static](https://github.com/koajs/static/tree/next) has been replaced by [koa-better-static](https://github.com/ohomer/koa-better-static)
- [koa-cors](https://github.com/evert0n/koa-cors/) has been replaced by [kcors](https://github.com/koajs/cors) (official package)
- [koa-session](https://github.com/koajs/session) has been replaced by [koa-generic-session](https://github.com/koajs/generic-session)

@dj-hedgehog I'll let you update the middlewares' list. Thanks for your time!

**Comment by dj-hedgehog** [comment]
@Aurelsicoko I updated the list.
Since you have been using Koa v2 already with Strapi. What are your experiences so far? When do you plan to switch?

**Comment by Aurelsicoko** [comment]
@dj-hedgehog Thank you so much! I can tell you that's Koa v2 could be a game changer in the next months. I'm very excited about the next release of Node without the harmony flag for async/await.

Here a summary of my feedback:

- Reducing the codebase (from 20% to 50% in some cases) thanks to async/await.
- The codebase is more readable and understandable.
- Better error handling. _(Btw in v3, we created a errors middleware based on top of Boom which automatically formats the errors and considerably reduces the number of try/catch patterns in your codebase)_
- I haven't notice performances improvements for now.

We should switch as soon as Node will allow using async/await without harmony flag. We are already working on big projects with `strapi@alpha.2`. It's so nice and most natural to develop an application like that.

**Comment by FabioSilvaBRZ** [comment]
Hi, guys.
I have been trying to install Strapi **V3** but I still have message erros like:
...
npm ERR! git rev-list -n1 v2.x: fatal: ambiguous argument 'v2.x': unknown revision or path not in the working tree.
npm ERR! git rev-list -n1 v2.x: Use '--' to separate paths from revisions, like this:
npm ERR! git rev-list -n1 v2.x: 'git <command> [<revision>...] -- [<file>...]'
...

I checked the packages been installed and the following situation came up:  in the error message it was trying to install like 29 packages - and I tried to install manually all of them to test if they work properly.
In a list of 25 packages there were a couple with the `.x` version.
`npm install "koa-bodyparser@koajs/bodyparser#3.x"`
`npm install "koa-send@koajs/send#2.x"`
They were installed just fine...

But there were 4 packages that do not install:
`npm install "koa-compress@koajs/compress#v2.x"`
`npm install "koa-favicon@koajs/favicon#v2.x"`
`npm install "koa-response-time@koajs/response-time#v2.x"`
`npm install "koa-mount@koajs/mount#next"`
They return the error reported above

Checking the GITHUB for these versions, one `.x` version that works I can find it at Github. For example: `"koa-bodyparser@koajs/bodyparser#3.x"`
![image](https://cloud.githubusercontent.com/assets/22884467/23456640/8c0680b4-fe74-11e6-996b-64f8ae3d4bf4.png)

But the `"koa-compress@koajs/compress#v2.x"` is not there, as follows:
![image](https://cloud.githubusercontent.com/assets/22884467/23456655/9deb887e-fe74-11e6-8a5e-c0e99954db1d.png)

Is it wrong? Long story short, those packages probably do not exist at all...

Another example? 
"koa-response-time@koajs/response-time#v2.x"
![image](https://cloud.githubusercontent.com/assets/22884467/23456674/b9d43900-fe74-11e6-99f8-4444968bb78f.png)
